cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/handlers/OpenScienceFrameworkInstitutionHandler.java

```diff
@@ -57,27 +72,41 @@ public String findInstitutionLogoutUrlById(final String id) {
     }
 
     /**
-     * Return a map of institution login url and institution name.
-     *  1.  The "name" is the value instead of the key.
-     *  2.  For institutions authenticated through "cas-pac4j", the institution id replaces the login url,
-     *      whose actual login url is generated during flow "client action".
+     * <p>Return a map of institution login url or institution ID as key and institution name as value.</p>
+     *
+     * <p>For saml-shib institutions, the login url is the key and the institution display name is the value.</p>
      *
-     * @param target The osf service target after successful institution login (only for "saml-shib" institutions)
-     * @return Map.
-     *      For "saml-shib", full login url as key and full institution display name as value;
-     *      For "cas-pac4j", institution id as key and full institution display name as value;
+     * <p>For cas-pac4j institutions, the institution ID is the key and the institution display name is the value. </p>
+     *
+     * @param target the osf service target after successful institution login, only used for "saml-shib" institutions
+     * @param id the OSF institution ID if auto-selection is enabled
+     * @return a single-entry {@link Map} if <code>id</code> presents and if the institution this <code>id</code>
+     *         identifies exists and supports institution SSO; otherwise return a multi-entry {@link Map} of all.
      */
-    public Map<String, String> getInstitutionLoginUrls(final String target) {
-        final List<OpenScienceFrameworkInstitution> institutionList = openScienceFrameworkDao.findAllInstitutions();
-        final Map<String, String> institutionLogin = new HashMap<>();
+    public Map<String, String> getInstitutionLoginUrlMap(final String target, final String id) {
+
+        List<OpenScienceFrameworkInstitution> institutionList = new LinkedList<>();
+        if (id == null || id.isEmpty()) {
+            institutionList = openScienceFrameworkDao.findAllInstitutions();
+        } else {
+            final OpenScienceFrameworkInstitution institution
+                    = openScienceFrameworkDao.findOneInstitutionById(id);
+            if (institution != null) {
+                institutionList.add(institution);
+            } else {
+                institutionList = openScienceFrameworkDao.findAllInstitutions();
+            }
+        }
```

Comment:
Now this handles both the one institution case and the all institutions one. If the pre-selected, institution is not found, back off to the all institutions case. The SSO eligibility is not needed here since the `institutionLoginUrlMap` building process below handles it.